### PR TITLE
[Refactor]: improve performance for detecting function components

### PR DESCRIPTION
**Pull request by golopot**
Benchmark on a private repo with `time npx eslint .` (real time):

|  | before | after |
| --- | -- | -- |
| run 1 | 6.6s | 4.8s |
| run 2 | 6.5s | 4.7s |
| run 3 | 6.7s | 4.8s |

Benchmark with `TIMING=10 npx eslint .`:

Before:
```
Rule                                | Time (ms) | Relative
:-----------------------------------|----------:|--------:
react/destructuring-assignment      |   345.923 |     7.0%
react/no-unstable-nested-components |   208.838 |     4.2%
react/no-did-mount-set-state        |   189.097 |     3.8%
react/boolean-prop-naming           |   187.663 |     3.8%
react/no-direct-mutation-state      |   186.664 |     3.8%
react/no-deprecated                 |   184.020 |     3.7%
react/hook-use-state                |   146.379 |     3.0%
react/no-access-state-in-setstate   |   145.210 |     3.0%
react/prefer-stateless-function     |   133.564 |     2.7%
react/require-optimization          |   128.301 |     2.6%
```

After:
```
Rule                                | Time (ms) | Relative
:-----------------------------------|----------:|--------:
react/no-did-mount-set-state        |   181.362 |     6.2%
react/destructuring-assignment      |   162.586 |     5.5%
react/no-deprecated                 |   134.652 |     4.6%
react/no-direct-mutation-state      |   124.110 |     4.2%
react/boolean-prop-naming           |   116.978 |     4.0%
react/no-did-update-set-state       |   113.393 |     3.9%
react/no-unknown-property           |    92.802 |     3.2%
react/hook-use-state                |    91.712 |     3.1%
react/no-unstable-nested-components |    90.477 |     3.1%
react/no-access-state-in-setstate   |    87.266 |     3.0%
```

### CODE REVIEW

File: lib/util/ast.js [hunk 1]
```diff
@@ -111,15 +115,23 @@ function traverseReturns(ASTNode, context, enterFunc) {
...
+        case 'IfStatement':
+        case 'ForStatement':
+        case 'WhileStatement':
+        case 'SwitchStatement':
+        case 'SwitchCase':
+          return;
         default:
+          this.skip();
       }
     },
```
Comment (golopot) [inline lib/util/ast.js:136]: Avoids unnecessary traversals by skipping all node types but the handled cases.

File: lib/util/jsx.js [hunk 2]
```diff
@@ -95,63 +94,48 @@ function isWhiteSpaces(value) {
...
+  };
 
-    return found && estraverse.VisitorOption.Break;
+  let found = false;
+  astUtil.traverseReturns(ASTnode, context, (node, breakTraverse) => {
+    if (isJSXValue(node)) {
+      found = true;
+      breakTraverse();
+    }
   });
```
Comment (golopot) [inline lib/util/jsx.js:139]: Same as above and simplify by not using `astUtil.traverse`.

File: tests/util/ast.js [hunk 3]
```diff
@@ -96,8 +96,8 @@ describe('ast', () => {
 
       assert.strictEqual(enterCalls.length, 6);
 
-      enterCalls.forEach((node, idx) => {
-        assert.strictEqual(node.lastArg.argument.value, idx);
+      enterCalls.forEach((call, idx) => {
+        assert.strictEqual(call.args[0].value, idx);
```
Comment (ljharb) [inline tests/util/ast.js:100]: why the change from last arg to first arg?

**Review by ljharb — COMMENTED** [review]
I can't pretend I understand this change, but if all the tests pass, then i guess yay?

File: tests/util/ast.js (hunk 3 above)
Comment (golopot) [reply tests/util/ast.js:100]: Previously the called function has only one parameter: `(node) => void`, I refactored that function to `(node, fn) => void`. The last arg changed to a different thing.

**Comment by golopot** [comment]
The gist is that `isReturningJSX` tries to find return statements that contains JSX. For that a tree traversing is performed. Previously the tree traversing is inefficient because it make unnecessary traversals to nodes that cannot contain a return statement. This PR makes the tree traversal more efficient by only traverse what is needed.

**Review by ljharb — APPROVED** [review]
